### ISSUE TITLE
Context manager for Threading

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#3245]
#### Describe the bug
It seems like hyperspy isn't doing a great job of managing multi-threading etc. for larger jobs. 
see https://github.com/pyxem/pyxem-demos/issues/87#issuecomment-1764825645 for some more information

#### Expected behavior
Ideally only `dask` would manage the number of threads for each action.  In that case we probably want to set some some of the optimal number of threads for different packages. i.e.

See https://docs.dask.org/en/stable/array-best-practices.html#avoid-oversubscribing-threads for more information

```
export OMP_NUM_THREADS=1
export MKL_NUM_THREADS=1
export OPENBLAS_NUM_THREADS=1
```

#### Additional context

Thanks to @uellue for bringing this up!  

I think many some sort of context manager would be a good way to handle this but @ericpre I was wondering if you had any additional ideas. 

I think we should just take the `map` function and then wrap the `map_blocks` function with a context manager similar to [LiberTEM's](https://github.com/LiberTEM/LiberTEM/blob/eeb7e71e43ceac28d02a886daf325a904978d7fd/src/libertem/common/threading.py#L198C5-L198C20). That should be sufficient....

**Comment by ericpre** [comment]
Yes, indeed, it should be facilitated where possible and this will depend on the library used. [`threadpoolctl` ](https://github.com/joblib/threadpoolctl/#setting-the-maximum-size-of-thread-pools) seems to support openmp and most blas libraries, so it is definitely worth considering it!

I can't think of a function in hyperspy which would be affected by this issue? However, it would be good to mentioned it explicitly in the documentation of `map`.

These resources are also worth a read:
https://scikit-learn.org/stable/computing/parallelism.html
https://thomasjpfan.github.io/pydata-nyc-2022-parallelism

numexpr can also be affected by this, but the way we used it, I don't think this is the case, maybe with samfire?

**Comment by CSSFrancis** [comment]
> Yes, indeed, it should be facilitated where possible and this will depend on the library used. [`threadpoolctl` ](https://github.com/joblib/threadpoolctl/#setting-the-maximum-size-of-thread-pools) seems to support openmp and most blas libraries, so it is definitely worth considering it!
> 
> I can't think of a function in hyperspy which would be affected by this issue? However, it would be good to mentioned it explicitly in the documentation of `map`.

Wouldn't almost all of the functions that use both `map` and some combination of `numba` or `scipy` run into the same problem?

For the `map` functions would it be useful to add a parameter such as `num_threads` for setting the number of threads equal to 1 for all cases?

> These resources are also worth a read: https://scikit-learn.org/stable/computing/parallelism.html https://thomasjpfan.github.io/pydata-nyc-2022-parallelism
> 
> numexpr can also be affected by this, but the way we used it, I don't think this is the case, maybe with samfire?

**Comment by ericpre** [comment]
> Wouldn't almost all of the functions that use both `map` and some combination of `numba` or `scipy` run into the same problem?

Yes, only if these function are using mutlithreading, which I don't think that there is - not something that I have checked to be honest!

> For the map functions would it be useful to add a parameter such as num_threads for setting the number of threads equal to 1 for all cases?

Yes, both approaches can be used, adding a `num_threads` argument to `map` to use `threadpool_limits(limits=1, user_api='blas')` context manager doesn't prevent setting threads externally through context manager or other. 
In any case, the user will need to set the number of thread explicitly, whether this is with a context manager or with `num_threads`. The advantage of adding `num_threads` to `map` is that the syntax is more concise but the drawback is that it is less transparent in what it does, however, if it should be able to explain clearly enough in the docstring and user guide what it does.
This sounds like that what matters most here is documenting when setting threading is relevant and sign posting to the relevant part of library documentation, such as scipy, numpy, numba, numexpr, etc to be able to figure out fairly quickly what function are using threading, etc.

**Comment by uellue** [comment]
Nice to hear that it can be useful! Just as a heads-up, detecting the libraries to control in `threadpool_limits` took a long time when we implemented it: https://github.com/LiberTEM/LiberTEM/issues/1117 That's why we implemented this mitigation that detects the libraries a few times, checks for changes in the loaded libraries, and then stops detecting once the set of libraries has stabilized:

https://github.com/LiberTEM/LiberTEM/blob/eeb7e71e43ceac28d02a886daf325a904978d7fd/src/libertem/common/threading.py#L163-L188

Not sure if threadpoolctl has improved this since then, just sth to watch out for. Feel free to use this!
